File: geral/livro_entendendo_algoritmos/quicksort/quicksort.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
)

func main() {
	values := load()

	quicksort(values, 0, len(values)-1)

	fmt.Println(values)
}

func load() []int {
	file, err := os.Open("unsorted.txt")

	if err != nil {
		log.Fatal(err)
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)

	var values []int

	for scanner.Scan() {
		value, _ := strconv.Atoi(scanner.Text())

		values = append(values, value)
	}

	return values
}

func quicksort(values []int, left, right int) {
	if left >= right {
		return
	}

	pivot := values[(left+right)/2]

	index := partition(values, left, right, pivot)

	quicksort(values, left, index-1)

	quicksort(values, index, right)
}

func partition(values []int, left, right, pivot int) int {
	for left <= right {
		for values[left] < pivot {
			left++
		}

		for values[right] > pivot {
			right--
		}

		if left <= right {
			swap(values, left, right)
			left++
			right--
		}
	}

	return left
}

func swap(values []int, i, j int) {
	values[i], values[j] = values[j], values[i]
}

File: geral/livro_entendendo_algoritmos/quick_sort.rb
def main()
    values = load

    quicksort(values, 0, values.size - 1)

    puts values
end

def load()
    values = []

    IO.foreach("unsorted.txt") { |line| values.push line.to_i }

    values
end

def quicksort(values, left, right)
    if left >= right
        return
    end

    pivot = values[(left + right) / 2]

    index = partition(values, left, right, pivot)

    quicksort(values, left, index - 1)

    quicksort(values, index, right)
end

def partition(values, left, right, pivot)
    while left <= right
        while values[left] < pivot
            left += 1
        end

        while values[right] > pivot
            right -= 1
        end

        if left <= right
            swap(values, left, right)
            left += 1
            right -= 1
        end
    end

    left
end

def swap(values, i, j)
    values[i], values[j] = values[j], values[i]
end

main
